Decouple from Buzz

fbourigault opened the pull request:
As discussed in #171 this replace the buzz client by [httplug](http://docs.php-http.org/en/latest/httplug/introduction.html).

I tried to implement this as it was done by @Nyholm for https://github.com/KnpLabs/php-github-api. This remove all the HttpClient stuff which was built on the top of Buzz. Instead it use Httplug standard plugins with a few custom ones for error handling and authentication.

I set PHP required version to 5.6+ as discussed in #171.

This **IS** a breaking change.

## To do
* [x] Update documentation.
* [x] Describe breaking changes.
* [x] Add tests.
* [x] Do intensive tests against real gitlab repositories.
* [x] Update ResultPager.
* [x] https://github.com/m4tthumphrey/php-gitlab-api/pull/191#issuecomment-308155979

Nyholm commented on lib/Gitlab/Api/AbstractApi.php:
```diff
@@ -1,6 +1,11 @@
 
 use Gitlab\Client;
+use Gitlab\HttpClient\Message\ResponseMediator;
+use function GuzzleHttp\Psr7\stream_for;
```
You need to remove this line. It will couple us to Guzzle. You should use the `StreamFactoryDiscovery` instead.

fbourigault commented:
@Nyholm I fixed the StreamFactory related issue. About the factory function, as there is many authentication methods, I don't know which one I should promote over the others.

I also added tests on some new classes, improved the `Gitlab\Client` code to limit the amount of breaking changes. I re-introduced the user-agent with a getter and a setter instead of the options array.

ResultPager is now fixed, but still waiting for unit tests.

Nyholm commented on lib/Gitlab/Client.php:
```diff
@@ -54,48 +58,62 @@ class Client
-        'user_agent'  => 'php-gitlab-api (http://github.com/m4tthumphrey/php-gitlab-api)',
-        'timeout'     => 60
-    );
+    private $userAgent = 'php-gitlab-api (http://github.com/m4tthumphrey/php-gitlab-api)';
```
Do we really need this in a property? We never read from it (as I can see)

fbourigault replied:
Users where able to read it through `getOption('user_agent')`. I added it to limit the breaking changes surface.

fbourigault commented:
I removed useless methods kept to limit breaking changes.
I documented the breaking changes in `UPGRADE.md` and also updated `README.md`.

fbourigault commented:
ResultPager is currently broken. Waiting for https://github.com/php-http/client-common/pull/74 and a release.

fbourigault commented:
I no longer require https://github.com/php-http/client-common/pull/74. I ended with a ApiVersion plugin which add the `/api/v3/` part when required.

This PR is definitively ready for review.

m1guelpf commented:
ping @m4tthumphrey

Nyholm commented on README.md:
````diff
@@ -7,25 +7,14 @@ Based on [php-github-api](https://github.com/m4tthumphrey/php-github-api) and co
-
-3. Include Composer's autoloader:
+```bash
+composer require m4tthumphrey/php-gitlab-api php-http/guzzle6-adapter
````
We should add php-http/message here. We could also add a link to http://docs.php-http.org/en/latest/httplug/users.html

fbourigault replied:
When using `php-http/guzzle6-adapter`, there is no need for `php-http/message`, right?

Nyholm replied:
ah, okey.

Nyholm commented on lib/Gitlab/Client.php:
```diff
@@ -54,48 +60,50 @@ class Client
+     *
+     * @return Client
+     */
+    public static function createWithHttpClient(HttpClient $httpClient)
```
Hm, I know I asked before. Maybe create a factory function for URL?

`Client::create($url)`?

fbourigault replied:
Looks interesting as many companies/users run their own GitLab instance.

Nyholm replied:
Yeah, exactly.

```php
$client = new \Gitlab\Client();
$client->setUrl('http://git.yourdomain.com');

// Could be the same as
$client = Gitlab\Client::create('http://git.yourdomain.com');
```

Nyholm commented on lib/Gitlab/Client.php:
```diff
@@ -250,50 +243,34 @@ public function clearHeaders()
+     * @return AbstractApi
      */
-    public function getOption($name)
+    public function __get($api)
```
FYI, https://github.com/mailgun/mailgun-php/pull/192#discussion_r84417626

We decided to remove it in Mailgun api

fbourigault replied:
Humm. This should not be in the diff. I will give it a look. Moreover, I also prefer writing methods and remove the magic method and huge comment in the top of the file. It's the same amount of work!

fbourigault replied:
And this will be part of an other PR.

Nyholm commented on lib/Gitlab/HttpClient/Plugin/ApiVersion.php:
```diff
@@ -0,0 +1,29 @@
+    {
+        $uri = $request->getUri();
+
+        if (substr($uri->getPath(), 0, 8) !== '/api/v3/') {
```
Hm, But what if there get new endpoints with `/api/v4/`?

fbourigault replied:
There are no plans to support multiple versions in a single release. V3 api will be discontinued in august, so the plan is to replace this by v4.

fbourigault commented:
What about removing `Client::setUserAgent` and documenting how to customize it like https://github.com/KnpLabs/php-github-api/blob/master/doc/customize.md#configure-the-http-client?

Nyholm commented:
👍

fbourigault commented:
As we are breaking, I cleaned  a bit more the `Gitlab\Client` public API. I will update `UPGRADE.md` and add a `doc/customize.md` to document how to set the client timeout, how to customize the user-agent and how to add custom headers.

fbourigault commented:
Documentation is now up-to-date.

m1guelpf commented:
@fbourigault Is this ready now?

fbourigault commented:
Yes! I will change things only if a reviewer can convince me to do so!

Nyholm commented on UPGRADE.md:
```diff
@@ -0,0 +1,14 @@
+
+## `Gitlab\Client` changes
+
+* The constructor no longer allow to specify base url. Use `setUrl` instead.
```
Use `setUrl` or `Client::create` instead.

Nyholm commented on lib/Gitlab/Client.php:
```diff
@@ -54,48 +60,63 @@ class Client
+
+        $this->httpClientBuilder->addPlugin(new GitlabExceptionThrower());
+        $this->httpClientBuilder->addPlugin(new HistoryPlugin($this->responseHistory));
+        $this->httpClientBuilder->addPlugin(new RedirectPlugin());
```
Do we really need this?

Nyholm commented on lib/Gitlab/HttpClient/Message/ResponseMediator.php:
```diff
@@ -0,0 +1,65 @@
+namespace Gitlab\HttpClient\Message;
+
+use Psr\Http\Message\ResponseInterface;
+
```
Maybe a class comment?

fbourigault commented:
@m4tthumphrey, do you agree with these changes?

fbourigault commented:
Is it ok if I move forward with next release and merge this PR by the end of this week? ping @m4tthumphrey @jubianchi @radutopala

fbourigault commented:
https://twitter.com/m4tthumphrey/status/883245976198275072

m1guelpf commented:
@fbourigault Did you tag a new release?

fbourigault commented:
Not yet, we have to make the library v4 only.

m1guelpf commented:
@fbourigault Do you have a task list/milestone?

fbourigault commented:
I will post a task list today with some "instructions" to be efficient.

m1guelpf commented:
@fbourigault Please ping me then :smile:

m4tthumphrey commented:
Amazing @fbourigault! Top work!
